sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala

```diff
@@ -0,0 +1,176 @@
+ * block1 = [sb11, sb12, sb13] // where sb13 is smaller than window length
+ * block2 = [b2]
+ *
+ * window.next() // sb11
+ * window.next() // sb12
+ * window.next() // [sb13, b2]
+ *
+ * @param blocks - sequence of blocks to manage
+ * @param windowSize - the size (in bytes) that block ranges should fit
+ * @tparam T - the specific type of `BlockWithSize`
+ * @note this class does not own `transferBlocks`
+ * @note this class is not thread safe
+ */
+class WindowedBlockIterator[T <: BlockWithSize](blocks: Seq[T], windowSize: Long)
+    extends Iterator[Seq[BlockRange[T]]] {
+
+  require(windowSize > 0, s"Invalid window size specified $windowSize")
+
+  private case class BlockWindow(start: Long, size: Long) {
+    val end = start + size // exclusive end offset
+    def move(): BlockWindow = {
+      BlockWindow(start + size, size)
+    }
+  }
+
+  // start the window at byte 0
+  private[this] var window = BlockWindow(0, windowSize)
+  private[this] var done = false
+
+  // helper class that captures the start/end byte offset
+  // for `block` on creation
+  private case class BlockWithOffset[T <: BlockWithSize](
+      block: T, startOffset: Long, endOffset: Long)
+
+  private[this] val blocksWithOffsets = {
+    var lastOffset = 0L
+    blocks.map { block =>
+      require(block.size > 0, "Invalid 0-byte block")
+      val startOffset = lastOffset
+      val endOffset = startOffset + block.size
+      lastOffset = endOffset // for next block
+      BlockWithOffset(block, startOffset, endOffset)
+    }
+  }
+
+  case class BlocksForWindow(lastBlockIndex: Option[Int],
+      blockRanges: Seq[BlockRange[T]],
+      hasMoreBlocks: Boolean)
+
+  private def getBlocksForWindow(
+      window: BlockWindow,
+      startingBlock: Int = 0): BlocksForWindow = {
+    val blockRangesInWindow = new ArrayBuffer[BlockRange[T]]()
+    var continue = true
+    var thisBlock = startingBlock
+    var lastBlockIndex: Option[Int] = None
+    while (continue && thisBlock < blocksWithOffsets.size) {
+      val b = blocksWithOffsets(thisBlock)
+      // if at least 1 byte fits within the window, this block should be included
+      if (window.start < b.endOffset && window.end > b.startOffset) {
+        var rangeStart = window.start - b.startOffset
+        if (rangeStart < 0) {
+          rangeStart = 0
+        }
+        var rangeEnd = window.end - b.startOffset
+        if (window.end >= b.endOffset) {
+          rangeEnd = b.endOffset - b.startOffset
+        }
+        blockRangesInWindow.append(BlockRange[T](b.block, rangeStart, rangeEnd))
+        lastBlockIndex = Some(thisBlock)
+      } else {
+        // skip this block, unless it's before our window starts
+        continue = b.endOffset <= window.start
+      }
+      thisBlock = thisBlock + 1
+    }
+    val lastBlock = blockRangesInWindow.last
+    BlocksForWindow(lastBlockIndex,
+      blockRangesInWindow,
+      !continue || !lastBlock.isComplete())
+  }
+
+  private var lastSeenBlock = 0
```

Comment:
Nit: Would like to see this at the top of the class next to the other member variables and marked `private[this]` for consistency with those.

Reply:
Sure thing: https://github.com/NVIDIA/spark-rapids/pull/934/commits/9a5c6d790361a25f4f9e4f146803975287f522df